Fix file removal

mbunkus opened the pull request:
### What does this PR do?

It fixes the `file.remove` module function and `file.symlink` state function to handle removal of existing directory entry types other than files, symlinks or directories: both files failed when the existing entry was e.g. a socket or a FIFO.

### What issues does this PR fix or reference?

Both together fix #35621.

### Previous Behavior

Before the fix the `file.remove` module function did nothing and returned `False` when asked to remove an existing socket or FIFO.

The `file.symlink` state function returned an error that the entry already exists when asked to replace an existing socket or FIFO with a symlink, even if `force=True` was set. The function did not attempt to remove the entry.

### New Behavior

Both functions will now handle all existing entries: directories are still removed recursively whereas other existing entries are removed via `os.remove()` no matter their type.

### Tests written?

No

### Commits signed with GPG?

Yes

mbunkus committed:
File module: make file.remove with other types (FIFOs, sockets)

The current checks only take regular files, symlinks and directories
into account. For all other types (FIFOs, sockets) the module doesn't
even attempt to remove them.

The new code only special-cases directories. For all other types of
existing file system entries a simple os.remove() is attempted.

This is the first part of fixing #35621.

mbunkus committed:
File state: make symlink deal with other existing types (FIFOs, sockets)

The current checks only take regular files, symlinks and directories
into account. For all other types (FIFOs, sockets) the state doesn't
even attempt to remove them; it simply returns an error that something
exists.

The new code keeps handling of symlinks as-is while generalizing all
other cases by only checking if they exist, and if they do, letting
the 'file.remove' module function handle the removal if 'force' is
true.

The 'file.remove' module function suffered from the same problem. It
was fixed to handle all types in the previous commit.

This is the second part of the fix for #35621.

cachedout reviewed (changes requested):
The logic here looks very sound but I think this behavior should be documented in a changelog entry before this is merged. It's possible that somebody was relying on the old (incorrect) behavior.

mbunkus commented:
I guess you mean adding such a `.. versionchanged:` section in the doc section of both functions? Sure, I can do that. Which version number should I put there?

twangboy commented:
@mbunkus Neon

mbunkus commented:
I've added `versionchanged` entries to the docs of both files, re-based on today's `develop` branch & force-pushed. Please re-review. Thanks :)

mbunkus commented:
Oh, wait. I'll take care of the tests, too.

mbunkus commented:
I've fixed the logic & force-pushed. For me `integration.modules.test_file` and
`integration.states.test_file` pass now.

However, ` unit.states.test_file.TestFileState.test_symlink` still fails with:

```
Traceback (most recent call last):
  File "/net/sweet-chili/home/mosu/prog/salt/tests/unit/states/test_file.py", line 298, in test_symlink
    backupname='SALT'),
  File "/net/sweet-chili/home/mosu/prog/salt/salt/states/file.py", line 1700, in symlink
    __salt__['file.symlink'](target, name)
KeyError: u'file.symlink'
```

I can't make heads or tails of it. Can someone please help with with why that happens? It seems like the `file` module could not be loaded… but there isn't really any other output that could shed some light on the problem.

garethgreenaway commented:
@mbunkus Updating your branch to get the tests to rerun, hopefully it was just a fluke.  Can you also include a note in the release notes here (https://github.com/saltstack/salt/blob/neon/doc/topics/releases/neon.rst) about that change?  Thanks!

mbunkus committed:
update Neon release notes about file.symlink state/file.remove module

mbunkus commented:
I've rebased my changes on the current `develop` branch and added a commit adding comments to `neon.rst`.

Then I tried running the `unit.states.test_file`, and that one still fails with the aforementioned `key error`. Like I said above, not being a regular Salt developer I really have no idea why that happens, much less how to fix the issue.

Can someone more knowledgeable with Salt's test procedures than me **please** help with this? Thanks.

mbunkus commented:
Alright, I know what's happening & how to fix it.

mbunkus commented:
I've fixed the tests & force-pushed. On my machine both the state & the module tests succeed now. Hopefully they do on Jenkins, too.

mbunkus commented:
Sooooo… now that all tests have passed, any chance of getting this merged? Anything else I can do?

mbunkus commented:
Ping! Anything else I can do to finally get this merged?

mbunkus commented:
I have to say I'm very disappointed how this PR has been handled. This is a bug fix, not some obscure feature enhancement. I've invested quite a lot of time figuring everything out wrt. test setups, running tests & how mocks work without a lot of (if any) help from the Salt community. And even after getting everything ready and pinging multiple times the PR hasn't been merged, nor has there been any feedback what's needed in order to get it merged. In short: it has been completely disregarded.

The bug it fixes is 3.75 years old now!

The handling of this bug is one of the reasons that motivated me to move away from Salt, meaning I won't spend any more time getting this thing merged (or even updated so that it works again). If you want the fix, please do something about it soon. I'll probably remove my cloned Salt git repo including the fix in a month or two.

dwoz commented:
@mbunkus Sorry for the extremely slow turn-around on this one. We've been making some changes to our process which should start to help us work through the PR backlog more quickly.  I've rebased this to `master` branch and will keep an eye on it.

dwoz committed:
Fix merge wart

dwoz committed:
Fix docs
